File: server/src/main/java/org/codi/catan/model/game/Hand.java
/*
 * @author the-codinator
 * created on 2020/5/28
 */

package org.codi.catan.model.game;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import java.util.EnumMap;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.codi.catan.util.Util;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public class Hand {

    @JsonInclude(Include.NON_NULL)
    private EnumMap<Resource, Integer> resources;
    private List<DevCard> devCards;

    @JsonIgnore
    public int getResourceCount() {
        return Util.getFrequencyMapTotalCount(resources);
    }

    @JsonIgnore
    public int getDevCardCount() {
        return devCards == null ? 0 : devCards.size();
    }
}

File: server-azure/src/model/game/hand.ts
import type { DevCard } from './dev-card';
import type { Resource } from './resource';
import { getFrequencyMapTotalCount } from '../../util/util';

export interface Hand {
  resources: Partial<Record<Resource, number>>;
  devCards: DevCard[];
}

export function getResourceCount(hand: Hand): number {
  return getFrequencyMapTotalCount(hand.resources);
}
